### PR TITLE
Refactor BSONFormats to reduce object allocation

### LINKED ISSUE #97: High Object Allocation causing unnecessary garbage collection
### ReactiveMongo-Play-Json Version (0.11.6, 0.11.11 / etc)
0.12.7-play25

### ReactiveMongo Version (0.11.6, 0.11.11 / etc)
0.12.7

### MongoDB version (2.6, 3.0 / etc)
3.4.4

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)
Linux ip-172-31-27-188 4.4.0-1037-aws #40-Ubuntu SMP Thu Dec 6 11:23:54 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)
 java version "1.8.0_171"
Java(TM) SE Runtime Environment (build 1.8.0_171-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.171-b11, mixed mode)

### Library Dependencies
Playversion 2.5.6

### Expected Behavior
On profiling a read-heavy application, I noticed high number of gc events.
On further profiling using JFR, it showed `BSONFormats.toJson` and `BSONFormats.toJson` create lots of unnecessary `PartialFunction.OrElse` instances. 

Please describe the expected behavior of the issue, starting from the first action.

1. The converters for JSON <-> BSON conversion should be initialized once and reused.

### Actual Behavior

The converters are initialized for every serialization/deserialization operation [here](https://github.com/ReactiveMongo/ReactiveMongo-Play-Json/blob/d9820440988f11266ae4e63075d4e63a5271b432/src/main/scala/package.scala#L651).

screen shot 2019-03-08 at 10 38 39 pm

### PR DESCRIPTION
**Pull request by chetankm-cs**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://github.com/ReactiveMongo/ReactiveMongo/blob/master/CONTRIBUTING.md#reactivemongo-developer--contributor-guidelines)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/2.4.x/WorkingWithGit#Squashing-commits)?
* [ ] Have you added tests for any changed functionality?

## Fixes

Fixes unnecessary instance creation of `orElse` on every invocation of `BsonFormats.toJson` and `BsonFormats.fromJson` . 

## Purpose
Combines `partialReads` and `partialWrites` and re-uses the resulting instance on every invocation of `BsonFormats.toJson` and `BsonFormats.fromJson`. 

## References
Are there any relevant issues / PRs / mailing lists discussions?
Fixes #97

### CODE REVIEW

File: src/main/scala/package.scala [hunk 1]
```diff
@@ -602,13 +602,15 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
     case JsNumber(n)                   => JsSuccess(BSONLong(n.toLong))
   }
 
-  private val defaultRead: JsValue => JsResult[BSONValue] =
-    readAsBSONValue(_: JsValue)
+  private val defaultRead: JsValue => JsResult[BSONValue] = {
```
Comment (cchantep) [inline src/main/scala/package.scala]: Move `getOrElse(JsError(s"unhandled json value: $json")` there doesn't change anything about the laziness.

File: src/main/scala/package.scala [hunk 2]
```diff
@@ -602,13 +602,15 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
...
+    val reads = combinedReads
+    json => reads.lift(json).getOrElse(JsError(s"unhandled json value: $json"))
+  }
 
   def toBSON(json: JsValue): JsResult[BSONValue] = defaultRead(json)
 
   @SuppressWarnings(Array("MaxParameters"))
-  def readAsBSONValue(json: JsValue)(
+  def combinedReads(
```
Comment (cchantep) [inline src/main/scala/package.scala]: Rename is breaking the API compat

File: src/main/scala/package.scala [hunk 3]
```diff
@@ -628,7 +630,7 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
     array: PartialReads[BSONArray],
     doc: PartialReads[BSONDocument],
     undef: PartialReads[BSONUndefined.type]
-  ): JsResult[BSONValue] =
+  ): PartialFunction[JsValue, JsResult[BSONValue]] =
```
Comment (cchantep) [inline src/main/scala/package.scala]: Changing return type is breaking the API

File: src/main/scala/package.scala [hunk 4]
```diff
@@ -646,15 +648,19 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
       orElse(symbol.partialReads).
       orElse(array.partialReads).
       orElse(doc.partialReads).
-      orElse(undef.partialReads).
-      lift(json).getOrElse(JsError(s"unhandled json value: $json"))
+      orElse(undef.partialReads)
 
-  private val defaultWrite: BSONValue => JsValue = writeAsJsValue(_: BSONValue)
+  private val defaultWrite: BSONValue => JsValue = {
```
Comment (cchantep) [inline src/main/scala/package.scala]: Moving `throw new JSONException(s"Unhandled json value: $bson")` doesn't change the laziness

File: src/main/scala/package.scala [hunk 5]
```diff
@@ -646,15 +648,19 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
       orElse(symbol.partialReads).
       orElse(array.partialReads).
       orElse(doc.partialReads).
-      orElse(undef.partialReads).
-      lift(json).getOrElse(JsError(s"unhandled json value: $json"))
```
Comment (cchantep) [inline src/main/scala/package.scala]: Break the compat with the previous/expected behaviour

File: src/main/scala/package.scala [hunk 6]
```diff
@@ -691,10 +697,7 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
     orElse(symbol.partialWrites).
     orElse(array.partialWrites).
     orElse(doc.partialWrites).
-    orElse(undef.partialWrites).
-    lift(bson).getOrElse(
```
Comment (cchantep) [inline src/main/scala/package.scala]: Break the compat with the previous/expected behaviour

File: src/main/scala/package.scala [hunk 7]
```diff
@@ -674,7 +680,7 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
     array: PartialWrites[BSONArray],
     doc: PartialWrites[BSONDocument],
     undef: PartialWrites[BSONUndefined.type]
-  ): JsValue = string.partialWrites.
+  ): PartialFunction[BSONValue, JsValue] = string.partialWrites.
```
Comment (cchantep) [inline src/main/scala/package.scala]: Changing return type is breaking the API

File: src/main/scala/package.scala [hunk 8]
```diff
@@ -646,15 +648,19 @@ sealed trait BSONFormats extends LowerImplicitBSONHandlers {
...
+    val writes = combinedWrites
+    bson => writes.lift(bson).getOrElse {
+      throw new JSONException(s"Unhandled json value: $bson")
+    }
+  }
 
   def toJSON(bson: BSONValue): JsValue = defaultWrite(bson)
 
   @SuppressWarnings(Array("MaxParameters"))
-  def writeAsJsValue(bson: BSONValue)(
```
Comment (cchantep) [inline src/main/scala/package.scala]: Rename is breaking the API compat

**Comment by chetankm-cs** [comment]
@cchantep Thanks for reviewing, I will try to fix these issue and submit pull request again. 
I am trying to fix #97 issue.

**Comment by cchantep** [comment]
#97 is already closes with `defaultX`

**Comment by chetankm-cs** [comment]
> #97 is already closes with `defaultX`

**Comment by chetankm-cs** [comment]
> #97 is already closes with `defaultX`

@cchantep I know that you have closed the issue, but it is not fixed. 
I have given an example also explaining why it is not fixed.

**Pull request closed without merging**